### PR TITLE
feat: added option for using private network

### PR DESCRIPTION
**Pull request by ptr1120**
Added option to use Hetzner private network for docker-machine communication.
Using this option in e.g. Rancher makes it possible to restrict node communication to the private network (you can disable the public network interface). See also https://github.com/mxschmitt/ui-driver-hetzner/issues/44.

### CODE REVIEW

**Comment by mxschmitt** [comment]
> Added option to use Hetzner private network for docker-machine communication.
> Using this option in e.g. Rancher makes it possible to restrict node communication to the private network (you can disable the public network interface). See also [mxschmitt/ui-driver-hetzner#44](https://github.com/mxschmitt/ui-driver-hetzner/issues/44).

Looks great! Also thought and tried this approach as far as I remember in the past to fix the linked issue. But if it works that way, then its perfect!

File: driver.go [hunk 1]
```diff
@@ -205,6 +213,10 @@ func (d *Driver) PreCreateCheck() error {
 		return errors.Wrap(err, "could not get location")
 	}
 
+	if d.UsePrivateNetwork == true && (d.networks == nil || len(d.networks) <= 0) {
```
Comment (mxschmitt) [inline driver.go]:
```suggestion
	if d.UsePrivateNetwork == true && (d.networks == nil || len(d.networks) == 0) {
```
the length can't be smaller than 0 afaik.

File: driver.go [hunk 2]
```diff
@@ -319,9 +331,26 @@ func (d *Driver) Create() error {
...
 
-	log.Debugf(" -> Server %s[%d] ready", srv.Server.Name, srv.Server.ID)
-	d.IPAddress = srv.Server.PublicNet.IPv4.IP.String()
+	if d.UsePrivateNetwork {
+		for {
+			// we need to wait until network is attached
+			log.Infof("Wait until private network attached ...")
+			server, _, err := d.getClient().Server.GetByID(context.Background(), srv.Server.ID)
+			if err != nil {
+				return err
```
Comment (mxschmitt) [inline driver.go]: maybe wrap the error and provide some more verbose output?

**Comment by ptr1120** [comment]
How can I set an option in ui-driver for docker-machine? Is this possible?
I tried:

```
<div class="col-md-2">
          <div class="checkbox">
            <label class="acc-label">{{input type="checkbox" checked=model.hetznerConfig.UsePrivateNetwork}} Use private network</label>
          </div>
</div>
```
But this doesn't work. I would like to test this implementation a little bit more, but currently, I can not configure this by the UI (only in code), which makes it difficult to test different configurations. Additionally updating the driver in Rancher is not working as expected currently I have to overwrite the driver in rancher-data/management-state/machine-drivers/ manually and restart rider if I want to test a new version.

**Comment by mxschmitt** [comment]
> How can I set an option in ui-driver for docker-machine? Is this possible?
> I tried:
> 
> ```
> <div class="col-md-2">
>           <div class="checkbox">
>             <label class="acc-label">{{input type="checkbox" checked=model.hetznerConfig.UsePrivateNetwork}} Use private network</label>
>           </div>
> </div>
> ```
> 
> But this doesn't work. I would like to test this implementation a little bit more, but currently, I can not configure this by the UI (only in code), which makes it difficult to test different configurations. Additionally updating the driver in Rancher is not working as expected currently I have to overwrite the driver in rancher-data/management-state/machine-drivers/ manually and restart rider if I want to test a new version.

I'll figure it out, let me quickly setup a dev setup. Will provide the UI Driver integration soon once it's working.

**Comment by ptr1120** [comment]
> Looks great! Also thought and tried this approach as far as I remember in the past to fix the linked issue. But if it works that way, then its perfect!

I basically oriented my self on the other drivers (google, softlayer) at https://github.com/docker/machine. They also only have an option whether to use private or public Ip and set the driver `IpAdress `based on that.

File: driver.go [hunk 3]
```diff
@@ -205,6 +213,10 @@ func (d *Driver) PreCreateCheck() error {
 		return errors.Wrap(err, "could not get location")
 	}
 
+	if d.UsePrivateNetwork == true && (d.networks == nil || len(d.networks) == 0) {
```
Comment (JonasProgrammer) [inline driver.go]: Minor thing, but direct comparison with true/false is discouraged. Also `nil` slices have a defined length of 0, so `d.UsePrivateNetwork && len(d.networks) == 0` should be sufficient

Comment (ptr1120) [reply driver.go]: Nice, up to now no experiences with go on my side, but sounds reasonable.

**Comment by JonasProgrammer** [comment]
LGMT

**Comment by mxschmitt** [comment]
Version 2.1.0 is released: https://github.com/JonasProgrammer/docker-machine-driver-hetzner/releases/download/2.1.0/docker-machine-driver-hetzner_2.1.0_linux_amd64.tar.gz

Thanks @ptr1120 ui-driver update will follow in the next hour I guess.

**Pull request merged**